arrow-libs/fx/arrow-fx-coroutines/src/commonMain/kotlin/arrow/fx/coroutines/Racing.kt

```diff
@@ -0,0 +1,129 @@
+ * suspend fun winner(): String = racing {
+ *   race { delay(1000); "Winner" }
+ *   race { throw RuntimeException("Loser") }
+ * } // Winner (logged RuntimeException)
+ *
+ * suspend fun winner(): String = racing {
+ *   race { delay(1000); "loser" }
+ *   raceOrFail { throw RuntimeException("Loser") }
+ * } // RuntimeException
+ * ```
+ *
+ * **Important:** a racing program with no racers will hang forever.
+ * ```kotlin
+ * suspend fun never(): Nothing = racing { }
+ * ```
+ *
+ * @param handleException handle any exception that occurred in [RacingScope.race],
+ * by default it [Throwable.printStackTrace].
+ * @param block the body of the DSL that describes the racing logic
+ * @return the winning value of [A].
+ */
+public suspend fun <A> racing(
+  handleException: ((context: CoroutineContext, exception: Throwable) -> Unit)? = null,
+  block: RacingScope<A>.() -> Unit,
+): A = coroutineScope {
+  val exceptionHandler = handleException ?: defaultExceptionHandler()::handleException
+  select {
+    val scope = SelectRacingScope(this@select, this@coroutineScope, exceptionHandler)
+    block(scope)
+    // TODO add this check??
+    require(scope.racers.value.isNotEmpty()) { "A racing program with no racers can never yield a result." }
+  }
+}
+
+public interface RacingScope<A> {
+  public suspend fun race(
+    context: CoroutineContext = EmptyCoroutineContext,
+    block: suspend CoroutineScope.() -> A
+  )
+
+  public suspend fun raceOrFail(
+    context: CoroutineContext = EmptyCoroutineContext,
+    block: suspend CoroutineScope.() -> A
+  )
+}
+
+private class SelectRacingScope<A>(
+  private val select: SelectBuilder<A>,
+  private val scope: CoroutineScope,
+  private val handleException: (context: CoroutineContext, exception: Throwable) -> Unit
+) : RacingScope<A>, CoroutineScope by scope {
+  val racers: Atomic<List<Deferred<A>>> = Atomic(emptyList())
+
+  override suspend fun raceOrFail(
+    context: CoroutineContext,
+    block: suspend CoroutineScope.() -> A
+  ) {
+    /* First we create a lazy racer,
+     * and we add it in front of the existing racers such that we maintain correct order.
+     * After we've successfully registered the racer, we check for race conditions,
+     * and 'start' racing.
+     */
+    val racer = scope.async(
+      start = CoroutineStart.LAZY,
+      context = context,
+      block = block
+    )
+    racers.update { racer prependTo it }
+    if (isActive) {
+      require(racer.start()) { "Racer not started" }
+      return with(select) {
+        racer.onAwait.invoke(::identity)
+      }
+    }
+  }
+
+  // To not fail the entire race, we allow users to handle the exceptions but do not cancel the race.
+  @OptIn(DelicateRaiseApi::class)
+  override suspend fun race(context: CoroutineContext, block: suspend CoroutineScope.() -> A) =
+    raceOrFail {
+      try {
+        block()
+      } catch (e: RaiseCancellationException) {
+        // `Raise<E>` error is ignored... Can we do better here?
```

Comment:
> Maybe raise shouldn't be considered a failure? Perhaps raising in such a way is conceptually the same as returning a value. Alternatively, we can have a raceOrRaise

Hmm. That kind-of makes sense, but I'd still expect user to want to 'select' `E` and not `A` but supporting both would be great.

So how do you imagine the API:
 - race (first `A`)
 - raceOrRaise ( first `A` or `E`)
 - raceOrThrow (first `A`, `E` or `Throwable`)
 
 ?? Do we need something that 'selects' first `A` or `Throwable`?
 
Potentially we can do a different extension when you reside within `Raise`, so an extension `Raise<E>.racing(...)` but then it's probably still possible to easily select the non-Raise variant 🤔.

```kotlin
suspend fun Raise<E>.racing(...): A
suspend fun racing(..): A
```

This probably doesn't yield the benefit we'd want since the non-Raise `racing` would get suggested when `Raise` doesn't match or `Raise.racing` has an additional param.